### PR TITLE
Generate new access code

### PR DESCRIPTION
**Pull request by AdeHarrison**
# Motivation and Context
To generate a new access (IAC) code

# What has changed
Dropped case iac code column after migrating existing values from caseiacaudit table. IAC codes should only be stored/read (using latest) from this table.

# How to test?
test in conjunction with [rasrm-acceptance-tests PR](rasrm-acceptance-tests](https://github.com/ONSdigital/rasrm-acceptance-tests/pull/168)

delete all caseiacaudit records

start casesvc service on this branch, caseiacaudit table should be re-populated with iac values from case table and column now removed

"Generate new enrolment code" function from response-operations-ui should add a record to caseiacaudit table (only if no active iac code)

# Links
https://trello.com/c/WTjuaOv5/190-sus004-generate-a-new-access-code-iac-for-an-lms-case

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/casesvc/CaseSvcApplication.java [hunk 1]
```diff
@@ -231,4 +232,9 @@ public DistributedLockManager reportDistributedLockManager(RedissonClient rediss
         redissonClient,
         appConfig.getDataGrid().getReportLockTimeToLiveSeconds());
   }
+
+  @Bean
+  public Clock clock() {
```
Comment (atrozzo) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/CaseSvcApplication.java]: Not sure you need a singleton bean for such things . I Think you should just call it where you need it.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACService.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    saveNewIacForCase(casePk, iac);
+    auditCaseEvent(casePk);
+    return iac;
+  }
+
+  private void auditCaseEvent(Integer casePk) {
+    CaseEvent caseEvent =
+        CaseEvent.builder()
+            .category(CategoryDTO.CategoryName.GENERATE_ENROLMENT_CODE)
+            .createdBy(null) // TODO
```
Comment (atrozzo) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACService.java]: we should have a better comment on what is going on here, can you not just omit and let the quety save null value ?

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACService.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      final CaseEventRepository caseEventRepo,
+      final Clock clock) {
+    this.iacClient = iacClient;
+    this.caseIacAuditRepo = caseIacAuditRepo;
+    this.caseEventRepo = caseEventRepo;
+    this.clock = clock;
+  }
+
+  @Transactional
+  public String generateNewCaseIACCode(final Integer casePk) {
```
Comment (atrozzo) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACService.java]: is this method not doing too much ? Th iac looks as well should be part of the contract and passed from outside and not be part of the transactional level ( cosmetic )

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseIacAuditRepository.java [hunk 4]
```diff
@@ -2,8 +2,22 @@
...
+/** JPA Data Repository for CaseIacAudit table. */
 @Repository
-public interface CaseIacAuditRepository extends JpaRepository<CaseIacAudit, Integer> {}
+@Transactional(readOnly = true)
+public interface CaseIacAuditRepository extends JpaRepository<CaseIacAudit, Integer> {
+
+  /**
+   * Find the latest case iac
+   *
+   * @param caseFK
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseIacAuditRepository.java]: It's worth deleting the `@param` and `@return` if there is no more context to add

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java [hunk 5]
```diff
@@ -44,33 +44,6 @@
    */
   List<Case> findByCaseGroupId(UUID caseGroupFK);
 
-  /**
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java]: 👍

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java [hunk 6]
```diff
@@ -94,4 +67,12 @@
    * @return
    */
   Case findBySampleUnitId(@Param("sampleUnitId") UUID sampleUnitId);
+
+  /**
+   * Find a case by it's PK
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java]: It's worth deleting the @param and @return if there is no more context to add

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import uk.gov.ons.ctp.response.casesvc.endpoint.CaseIACEndpoint.CaseIACDTO;
+import uk.gov.ons.ctp.response.casesvc.message.notification.CaseNotification;
+import uk.gov.ons.ctp.response.casesvc.message.sampleunitnotification.SampleUnitParent;
+import uk.gov.ons.ctp.response.casesvc.representation.CaseDetailsDTO;
+import uk.gov.ons.tools.rabbit.Rabbitmq;
+import uk.gov.ons.tools.rabbit.SimpleMessageBase;
+import uk.gov.ons.tools.rabbit.SimpleMessageListener;
+import uk.gov.ons.tools.rabbit.SimpleMessageSender;
+
+@Slf4j
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: We should be using the GoDaddy logger

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import uk.gov.ons.ctp.common.endpoint.CTPEndpoint;
+import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.response.casesvc.domain.model.Case;
+import uk.gov.ons.ctp.response.casesvc.domain.model.CaseIacAudit;
+import uk.gov.ons.ctp.response.casesvc.service.CaseIACService;
+import uk.gov.ons.ctp.response.casesvc.service.CaseService;
+
+@RestController
+@RequestMapping(value = "/cases/{caseId}/iac", produces = "application/json")
+@Slf4j
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpoint.java]: We should be using the GoDaddy logger

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import uk.gov.ons.ctp.response.casesvc.domain.model.CaseEvent;
+import uk.gov.ons.ctp.response.casesvc.domain.model.CaseIacAudit;
+import uk.gov.ons.ctp.response.casesvc.domain.repository.CaseEventRepository;
+import uk.gov.ons.ctp.response.casesvc.domain.repository.CaseIacAuditRepository;
+import uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO;
+import uk.gov.ons.ctp.response.casesvc.service.CaseIACService;
+import uk.gov.ons.ctp.response.casesvc.service.InternetAccessCodeSvcClientService;
+
+@Service
+@Slf4j
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java]: We should be using the GoDaddy logger

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+   * @return a new SimpleMessageSender
+   */
+  private SimpleMessageSender getMessageSender() {
+    Rabbitmq config = this.appConfig.getRabbitmq();
+
+    return new SimpleMessageSender(
+        config.getHost(), config.getPort(), config.getUsername(), config.getPassword());
+  }
+
+  private String waitForNotification() throws Exception {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: Might be worth refactoring the common logic out

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+                    .withBody("[\"{{randomValue length=12 type='ALPHANUMERIC' lowercase=true}}\"]")
+                    .withTransformers("response-template")));
+  }
+
+  /**
+   * Creates a new SimpleMessageSender based on the config in AppConfig
+   *
+   * @return a new SimpleMessageSender
+   */
+  private SimpleMessageSender getMessageSender() {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: Might be worth refactoring the common logic out

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+    return message;
+  }
+
+  /**
+   * Creates a new SimpleMessageListener based on the config in AppConfig
+   *
+   * @return a new SimpleMessageListener
+   */
+  private SimpleMessageListener getMessageListener() {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: Might be worth refactoring the common logic out

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        .asString();
+  }
+
+  /*
+   * Sends a sample unit in a message so that casesvc creates a case, then waits for a message on
+   * the case lifecycle queue which confirms case creation
+   *
+   * @return a new CaseNotification
+   */
+  private CaseNotification sendSampleUnit(
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: Might be worth refactoring the common logic out

**Comment by AdeHarrison** [comment]
[SUS004: Generate a new access code (IAC) for an LMS case](https://trello.com/c/WTjuaOv5/190-sus004-generate-a-new-access-code-iac-for-an-lms-case)

**Comment by AdeHarrison** [comment]
[SUS004: Generate a new access code (IAC) for an LMS case](https://trello.com/c/WTjuaOv5/190-sus004-generate-a-new-access-code-iac-for-an-lms-case)

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseIacAuditRepository.java (hunk 4 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseIacAuditRepository.java]: fixed

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java (hunk 6 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java]: fixed

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpoint.java (hunk 8 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpoint.java]: fixed

File: src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java (hunk 7 above)
Comment (AdeHarrison) [reply src/test/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseIACEndpointIT.java]: fixed

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java (hunk 9 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java]: fixed

File: src/main/resources/database/changes/release-13/add_H_cases_to_OFFLINE_RESPONSE_PROCESSED_category.sql [hunk 14]
```diff
@@ -1,3 +1,3 @@
-UPDATE casesvc.category
```
Comment (benjefferies) [inline src/main/resources/database/changes/release-13/add_H_cases_to_OFFLINE_RESPONSE_PROCESSED_category.sql]: We should revert the styling changes

Comment (AdeHarrison) [reply src/main/resources/database/changes/release-13/add_H_cases_to_OFFLINE_RESPONSE_PROCESSED_category.sql]: fixed

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    saveNewIacForCase(casePk, iac);
+    auditCaseEvent(casePk);
+    return iac;
+  }
+
+  private void auditCaseEvent(Integer casePk) {
+    CaseEvent caseEvent =
+        CaseEvent.builder()
+            .category(CategoryDTO.CategoryName.GENERATE_ENROLMENT_CODE)
+            .createdBy(null) // TODO
```
Comment (emilioward) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java]: Why is this TODO?

Comment (benjefferies) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseIACServiceImpl.java]: Maybe this line can just be deleted. It should be set to the user doing the action but we don't have that knowledge

**Comment by emilioward** [comment]
Me and @liamtoozer have ran through them again with no errors so they don't appear to pick up this bug

**Pull request merged**